### PR TITLE
Unit Test - Test Share View Model

### PR DESCRIPTION
**Pull request by at-datbui**
#### Related issues (Required)
- Task #193 

#### Description (Required)
- [x] Unit test for share view model

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 1]
```diff
@@ -50,15 +52,24 @@ class WayRepository : WayDataSource {
...
     override fun getListLocation(url: String): Observable<ResultRoad> {
         return remoteDataSource.getListLocation(url)
     }
+
     override fun createAndAssignAction(builder: ActionParamsBuilder): Observable<Action>
             = remoteDataSource.createAndAssignAction(builder)
+
+    override fun getTrackingURL(): Single<String> = remoteDataSource.getTrackingURL()
+
+    override fun getLocationName(context: Context, latLng: LatLng): Single<String> =
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/997523018c86063473495f71abbfc6b6d8aa160d/app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt#L67"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 2]
```diff
@@ -165,20 +172,66 @@ internal class WayRemoteDataSource : WayDataSource {
...
-                val throwable = Throwable(error.errorMessage)
+            override fun onError(errorResponse: ErrorResponse) {
+                val throwable = Throwable(errorResponse.errorMessage)
+                link.onError(throwable)
+            }
+        })
+        return link
+    }
+
+
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/997523018c86063473495f71abbfc6b6d8aa160d/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L196"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 3]
```diff
@@ -165,20 +172,66 @@ internal class WayRemoteDataSource : WayDataSource {
...
+        }
+        return locationName
+    }
+
+    override fun getCurrentLocationHyperTrack(): Single<HyperTrackLocation> {
+        val result = SingleSubject.create<HyperTrackLocation>()
+        HyperTrack.getCurrentLocation(object : HyperTrackCallback() {
+            override fun onSuccess(p0: SuccessResponse) {
+                try {
+                    val hyperTrackLocation = HyperTrackLocation((p0.responseObject) as Location?)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/997523018c86063473495f71abbfc6b6d8aa160d/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L215"></span>Cast operator throws an exception if the cast is not possible.</td>
    </tr>
  </tbody>
</table>

**Comment by at-dev-support-01** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">All test passed. (32/32)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 4]
```diff
@@ -50,15 +52,23 @@ class WayRepository : WayDataSource {
...
     }
+
     override fun createAndAssignAction(builder: ActionParamsBuilder): Observable<Action>
             = remoteDataSource.createAndAssignAction(builder)
+
+    override fun getTrackingURL(): Single<String> = remoteDataSource.getTrackingURL()
+
+    override fun getLocationName(context: Context, latLng: LatLng): Single<String> = remoteDataSource.getLocationName(context, latLng)
+
+    override fun getCurrentLocationHyperTrack(): Single<HyperTrackLocation> =
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/ca73e64d6b6bbf21d29148d06ca6d0f051e320c5/app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt#L69"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 5]
```diff
@@ -165,20 +172,65 @@ internal class WayRemoteDataSource : WayDataSource {
...
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/ca73e64d6b6bbf21d29148d06ca6d0f051e320c5/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L214"></span>Cast operator throws an exception if the cast is not possible.</td>
    </tr>
  </tbody>
</table>

File: app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        /* Given */
+        val location = Location("point")
+        val resultTest = TestObserver<Location>()
+        `when`(wayRepository.getCurrentLocation(context)).thenReturn(Single.just(location))
+
+        /* When */
+        viewModel.getCurrentLocation(context).subscribe(resultTest)
+
+        /* Then */
+        resultTest.assertValue { true }
```
Comment (at-vinhhuynh) [inline app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt]: remove this line

File: app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        val test = TestObserver<HyperTrackLocation>()
+        val hyperTrackLocation = HyperTrackLocation(Location("point"))
+        `when`(wayRepository.getCurrentLocationHyperTrack()).thenReturn(Single.just(hyperTrackLocation))
+
+        /* When */
+        viewModel.getCurrentLocationHyperTrack().subscribe(test)
+
+        /* Then */
+        test.assertValue { true }
+        test.assertResult(hyperTrackLocation)
```
Comment (at-vinhhuynh) [inline app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt]: change this line to `test.assertResult(it == hyperTrackLocation)`

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 8]
```diff
@@ -131,12 +138,13 @@ internal class WayRemoteDataSource : WayDataSource {
...
-    override fun getCurrentLocation(): Observable<HyperTrackLocation> {
-        val result = AsyncSubject.create<HyperTrackLocation>()
-        HyperTrack.getCurrentLocation(object : HyperTrackCallback() {
+    override fun getETA(destination: LatLng, vehicle: VehicleType): Observable<Float> {
+        val result = AsyncSubject.create<Float>()
+        HyperTrack.getETA(destination, vehicle, object : HyperTrackCallback() {
             override fun onSuccess(response: SuccessResponse) {
-                val res = HyperTrackLocation((response.responseObject) as? Location?)
-                result.onNext(res)
+                response.responseObject.let {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/5bbe9fd97d250f332b004ef1e3f897a9e565edce/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L145"></span>Cast operator throws an exception if the cast is not possible.</td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 9]
```diff
@@ -165,20 +173,66 @@ internal class WayRemoteDataSource : WayDataSource {
...
+        return locationName
+    }
+
+    override fun getCurrentLocationHyperTrack(): Single<HyperTrackLocation> {
+        val result = SingleSubject.create<HyperTrackLocation>()
+        HyperTrack.getCurrentLocation(object : HyperTrackCallback() {
+            override fun onSuccess(p0: SuccessResponse) {
+                try {
+                    p0.responseObject.let {
+                        HyperTrackLocation(it as Location?)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/5bbe9fd97d250f332b004ef1e3f897a9e565edce/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L214"></span>Cast operator throws an exception if the cast is not possible.</td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 10]
```diff
@@ -165,20 +173,66 @@ internal class WayRemoteDataSource : WayDataSource {
...
+        if (addresses.isNotEmpty()) {
+            val address: Address = addresses[0]
+            locationName.onSuccess(address.getAddressLine(0))
+        } else {
+            locationName.onSuccess("")
+        }
+        return locationName
+    }
+
+    override fun getCurrentLocationHyperTrack(): Single<HyperTrackLocation> {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: change method name to `getCurrentHyperTrackLocation`

File: app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+        /* When */
+        viewModel.getCalendarDateTime().subscribe(test)
+
+        /* Then */
+        test.assertValue { it == simple.format(calendar) }
+    }
+
+    @Test
+    fun `Given a latLng point - When compare two point - Then return compare result`() {
```
Comment (at-vinhhuynh) [inline app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt]: add a method to test case return `false`

File: app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        val endLatLng = LatLng(16.0804405, 108.238518)
+        val startLatLng1 = LatLng(16.0804405, 108.238518)
+        val endLatLng1 = LatLng(16.0804426, 108.2385165)
+
+        /* When */
+        viewModel.getAngleMarker(startLatLng, endLatLng).subscribe(resultTest)
+        viewModel.getAngleMarker(startLatLng1, endLatLng1).subscribe(resultTest)
+
+        /* Then */
+        resultTest.assertValueAt(0, 0.0f)
```
Comment (at-vinhhuynh) [inline app/src/test/java/vn/asiantech/way/share/ShareViewModelTest.kt]: remove this 2 lines and replace by `resultTest.assertValues(0.0f, 325.53668f)`

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 13]
```diff
@@ -50,15 +52,23 @@ class WayRepository : WayDataSource {
...
     }
+
     override fun createAndAssignAction(builder: ActionParamsBuilder): Observable<Action>
             = remoteDataSource.createAndAssignAction(builder)
+
+    override fun getTrackingURL(): Single<String> = remoteDataSource.getTrackingURL()
+
+    override fun getLocationName(context: Context, latLng: LatLng): Single<String> = remoteDataSource.getLocationName(context, latLng)
+
+    override fun getCurrentHyperTrackLocation(): Single<HyperTrackLocation> =
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/314f3ff69a1edfdbe8485615a757ca98224a7714/app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt#L69"></span></td>
    </tr>
  </tbody>
</table>

**Comment by at-vinhhuynh** [comment]
@at-tienhoang 
I finish review this PR, please check it again.
thank you!

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 13 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/9954a62a32b2ad3519099e66ffb19aeeacc42205/app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt#L69"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 14]
```diff
@@ -131,12 +137,13 @@ internal class WayRemoteDataSource : WayDataSource {
...
-    override fun getCurrentLocation(): Observable<HyperTrackLocation> {
-        val result = AsyncSubject.create<HyperTrackLocation>()
-        HyperTrack.getCurrentLocation(object : HyperTrackCallback() {
+    override fun getETA(destination: LatLng, vehicle: VehicleType): Observable<Float> {
+        val result = AsyncSubject.create<Float>()
+        HyperTrack.getETA(destination, vehicle, object : HyperTrackCallback() {
             override fun onSuccess(response: SuccessResponse) {
-                val res = HyperTrackLocation((response.responseObject) as? Location?)
-                result.onNext(res)
+                response.responseObject.let {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/9954a62a32b2ad3519099e66ffb19aeeacc42205/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L144"></span>Cast operator throws an exception if the cast is not possible.</td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 15]
```diff
@@ -165,20 +172,67 @@ internal class WayRemoteDataSource : WayDataSource {
...
+        return locationName
+    }
+
+    override fun getCurrentHyperTrackLocation(): Single<HyperTrackLocation> {
+        val result = SingleSubject.create<HyperTrackLocation>()
+        HyperTrack.getCurrentLocation(object : HyperTrackCallback() {
+            override fun onSuccess(p0: SuccessResponse) {
+                try {
+                    p0.responseObject.let {
+                        HyperTrackLocation(it as Location?)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/9954a62a32b2ad3519099e66ffb19aeeacc42205/app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt#L213"></span>Cast operator throws an exception if the cast is not possible.</td>
    </tr>
  </tbody>
</table>

**Pull request merged**